nodebuilder/pruner/pruner.go

```diff
@@ -0,0 +1,308 @@
+		// exchange. Avoiding duplication on storage level is more expensive.
+		if _, ok := alreadyPruned[string(dh)]; ok {
+			continue
+		}
+		err = sp.store.Remove(ctx, dh)
+		alreadyPruned[string(dh)] = struct{}{}
+		if err != nil {
+			log.Errorw("failed to remove datahash", "datahash", dh, "epoch", epoch, "err", err)
+			continue
+		}
+		sp.metrics.observePrune(ctx)
+	}
+
+	key := datastore.NewKey(fmt.Sprintf("%d", epoch))
+	err = sp.ds.Delete(ctx, key)
+	if err != nil {
+		return fmt.Errorf("failed to delete epoch %d from datastore: %w", epoch, err)
+	}
+
+	delete(sp.activeEpochs, epoch)
+	if epoch == sp.oldestEpoch.Load() {
+		sp.unsafeUpdateOldestEpoch()
+	}
+	return nil
+}
+
+func (sp *StoragePruner) unsafeUpdateOldestEpoch() {
+	oldest := sp.oldestEpoch.Load()
+	next := oldest
+	for {
+		nextTime := sp.epochToTime(next).Add(sp.cfg.EpochDuration)
+		if nextTime.After(time.Now()) {
+			log.Warn("could not find new oldest epoch")
+			// set to sentinel
+			sp.oldestEpoch.CompareAndSwap(oldest, ^uint64(0))
+			break
+		}
+		nextEpoch := sp.calculateEpoch(nextTime)
+		_, ok := sp.activeEpochs[nextEpoch]
+		if ok {
+			sp.oldestEpoch.CompareAndSwap(oldest, nextEpoch)
+			break
+		}
+		next++
+	}
+}
+
+func (sp *StoragePruner) getDatahashesFromEpoch(ctx context.Context, epoch uint64) ([]share.DataHash, error) {
+	var datahashes []share.DataHash
+	key := datastore.NewKey(fmt.Sprintf("%d", epoch))
+	val, err := sp.ds.Get(ctx, key)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get datahashes for epoch %d: %w", epoch, err)
+	}
+	buf := bytes.NewBuffer(val)
+	dec := gob.NewDecoder(buf)
+	if err := dec.Decode(&datahashes); err != nil {
+		return nil, fmt.Errorf("failed to decode datahashes for epoch %d: %w", epoch, err)
+	}
+	return datahashes, nil
+}
+
+func (sp *StoragePruner) saveDatahashesToEpoch(ctx context.Context, epoch uint64, datahashes []share.DataHash) error {
+	key := datastore.NewKey(fmt.Sprintf("%d", epoch))
+
+	var buf bytes.Buffer
+	enc := gob.NewEncoder(&buf)
+	err := enc.Encode(datahashes)
+	if err != nil {
+		return fmt.Errorf("failed to marshal datahashes for epoch %d: %w", epoch, err)
+	}
+	err = sp.ds.Put(ctx, key, buf.Bytes())
+	if err != nil {
+		return fmt.Errorf("failed to put datahashes for epoch %d: %w", epoch, err)
+	}
+	return nil
+}
+
+func (sp *StoragePruner) globalLock() {
+	for i := range sp.stripedLocks {
+		sp.stripedLocks[i].Lock()
+	}
+}
+
+func (sp *StoragePruner) globalUnlock() {
+	for i := range sp.stripedLocks {
+		sp.stripedLocks[i].Unlock()
+	}
+}
+
+func (sp *StoragePruner) calculateEpoch(timestamp time.Time) uint64 {
+	return uint64(timestamp.Unix() / int64(sp.cfg.EpochDuration.Seconds()))
+}
+
+func (sp *StoragePruner) epochToTime(epoch uint64) time.Time {
+	return time.Unix(int64(epoch*uint64(sp.cfg.EpochDuration.Seconds())), 0)
+}
+
+func (sp *StoragePruner) epochIsRecent(epoch uint64) bool {
+	return epoch >= sp.calculateEpoch(time.Now().Add(-sp.cfg.RecencyWindow))
```

Comment:
maybe we can eventually do this by # of active epochs.

Reply:
Number of epochs isnt going to tell you if an epoch is recent or old enough to be pruned. The only thing number of epochs helps with is determining whether GC should run or not

Reply:
The number of the epoch could help indicate whether it's old or not (e.g. smallest number in epoch)

I'm fine for this being the implementation currently and then we can think of moving away from time-based as it's a bit more brittle. I don't think switching the method for epoch determination would be super breaking.